temporal-serviceclient/src/main/java/io/temporal/internal/AbstractBackoffThrottler.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal;
+
+import io.grpc.Status;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * Used to throttle code execution in presence of failures using exponential backoff logic. The
+ * formula used to calculate the next sleep interval is:
+ *
+ * <p>
+ *
+ * <pre>
+ * min(pow(backoffCoefficient, failureCount - 1) * initialSleep, maxSleep);
+ * </pre>
+ */
+abstract class AbstractBackoffThrottler {
+
+  private final Duration regularInitialSleep;
+
+  private final Duration longInitialSleep;
+
+  private final Duration maxSleep;
+
+  private final double backoffCoefficient;
+
+  private final AtomicReference<ThrottlerState> lastFailure =
+      new AtomicReference<>(INITIAL_THROTTLER_STATE);
+
+  /**
+   * Construct an instance of the throttler.
+   *
+   * @param regularInitialSleep time to sleep on the first failure (assuming regular failures)
+   * @param longInitialSleep time to sleep on the first failure (for failures that require longer
+   *     wait time)
+   * @param maxSleep maximum time to sleep independently of number of failures
+   * @param backoffCoefficient coefficient used to calculate the next time to sleep.
+   */
+  public AbstractBackoffThrottler(
+      Duration regularInitialSleep,
+      Duration longInitialSleep,
+      @Nullable Duration maxSleep,
+      double backoffCoefficient) {
+    Objects.requireNonNull(regularInitialSleep, "regularInitialSleep");
+    Objects.requireNonNull(longInitialSleep, "longInitialSleep");
+    if (backoffCoefficient < 1.0) {
+      throw new IllegalArgumentException(
+          "backoff coefficient less than 1.0: " + backoffCoefficient);
+    }
+    this.regularInitialSleep = regularInitialSleep;
+    this.longInitialSleep = longInitialSleep;
+    this.maxSleep = maxSleep;
+    this.backoffCoefficient = backoffCoefficient;
+  }
+
+  long calculateSleepTime() {
+    ThrottlerState state = lastFailure.get();
+    if (state.failureCount == 0) return 0;
+    Duration effectiveInitialSleep =
+        (state.lastFailureCode == Status.Code.RESOURCE_EXHAUSTED)
+            ? longInitialSleep
+            : regularInitialSleep;
+    double sleepMillis =
```

Comment:
Do I understand correctly, that (n+1)th sleep may be shorter than nth sleep if the gRPC code changed?
If it's the case and it's by design - a good comment explaining why it's by design is needed. Otherwise it will look like a bug for the next reader.

Reply:
Yes, that is the expected behaviour, per consensus. I agree that it is somewhat counter intuitive and I'll add a comment making it clear that this is known and agreed. 

However, I won't try to explain why we settled on that behaviour rather than any other. TBH I'd have a hard time arguing this without falling in bike shedding, so the only rationales that stands are essentially that its what the server team asked for and that none of us as come up with a better proposition...